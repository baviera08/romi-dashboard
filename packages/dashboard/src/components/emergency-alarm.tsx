import { IconButton } from '@material-ui/core';
import React from 'react';
import { Alerts } from '../util/alerts';
import NotificationsActiveIcon from '@material-ui/icons/NotificationsActive';

interface EmergencyAlarmProps {
  onTurnOn?: () => void;
  onTurnOff?: () => void;
}

export const EmergencyAlarm = React.memo((props: EmergencyAlarmProps) => {
  const { onTurnOn, onTurnOff } = props;
  const [alarm, setAlarm] = React.useState(false);

  const handleAlarm = (): void => {
    if (alarm) {
      Alerts.verification({
        confirmCallback: () => {
          setAlarm(false);
          onTurnOff && onTurnOff();
        },
        body: `You're about to turn off the alarm! The robots will resume their tasks.`,
      });
    } else {
      Alerts.verification({
        confirmCallback: () => {
          setAlarm(true);
          onTurnOn && onTurnOn();
        },
        body: `You're about to fire an alarm! The robots will head to their nearest holding points. Once you accept this there is no turning back.`,
      });
    }
  };

  return (
    <IconButton id="alarm-btn" color={alarm ? 'secondary' : 'inherit'} onClick={handleAlarm}>
      <NotificationsActiveIcon />
    </IconButton>
  );
});
