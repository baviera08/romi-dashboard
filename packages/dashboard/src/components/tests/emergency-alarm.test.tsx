import { mount, shallow } from 'enzyme';
import React from 'react';
import { EmergencyAlarm } from '../emergency-alarm';
import { IconButton } from '@material-ui/core';
/**
 * We can't test alters from sweetalert2 here, so the complete functionality is tested
 * in an e2e test.
 */
describe('Emergency Alarm behavior', () => {
  let root: any;

  afterEach(() => {
    root.unmount();
    jest.clearAllMocks();
  });

  test('Renders correctly', () => {
    root = shallow(<EmergencyAlarm></EmergencyAlarm>);
    expect(root).toMatchSnapshot();
  });

  test('Alarm shows a inherit color Icon when is deactivated', () => {
    root = mount(<EmergencyAlarm></EmergencyAlarm>);
    expect(root.find(IconButton).prop('color')).toBe('inherit');
  });

  test('Alarm shows a red icon when is activated', () => {
    const stateSetter = jest.fn();
    jest
      .spyOn(React, 'useState')
      //Simulate that mode state value was set to 'true'
      .mockImplementation((stateValue: boolean) => [(stateValue = true), stateSetter]);
    root = mount(<EmergencyAlarm></EmergencyAlarm>);
    expect(root.find(IconButton).prop('color')).toBe('secondary');
  });
});
