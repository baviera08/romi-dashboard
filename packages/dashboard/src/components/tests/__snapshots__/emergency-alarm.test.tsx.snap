// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Emergency Alarm behavior Renders correctly 1`] = `
<WithStyles(ForwardRef(IconButton))
  color="inherit"
  id="alarm-btn"
  onClick={[Function]}
>
  <NotificationsActiveIcon />
</WithStyles(ForwardRef(IconButton))>
`;

exports[`Renders correctly 1`] = `
<WithStyles(ForwardRef(IconButton))
  color="secondary"
  id="alarm-btn"
  onClick={[Function]}
>
  <NotificationsActiveIcon />
</WithStyles(ForwardRef(IconButton))>
`;
